### ISSUE TITLE
Help starting the app

### ISSUE THREAD

**Issue by andarms** [david-janssen/kmonad#160]
Hi, I'm a new Linux user. I want to use this software to remap my keyboard. However, the documentation is not clear, at least for me, a new Linux user. So, I have a few questions that I hope anyone could help me with:
- how to install?  	
There are some instructions about manual compilation, but  I try to use the precompiled bin for Linux on the release page, and it doesn't work. I made it an executable, try to run with a .kbd file as an argument, and any of this works.

- how to identify my keyboard device?
I look up the /dev/input/by-id directory, and when my USB keyboard is connected, I can see the device file that I will need to use in the configuration kbd file. But I'm using a laptop. Where can  I find the id to use in the configuration? And how to use the same setting for the integrated keyboard and the USB keyboard?

**Comment by MaxGyver83** [comment]
> - how to install?  	
> There are some instructions about manual compilation, but  I try to use the precompiled bin for Linux on the release page, and it doesn't work. I made it an executable, try to run with a .kbd file as an argument, and any of this works.

Sounds correct. Have you also seen the [FAQ](https://github.com/david-janssen/kmonad/blob/master/doc/faq.md#linux)? (uinput permissions)

> - how to identify my keyboard device?
> I look up the /dev/input/by-id directory, and when my USB keyboard is connected, I can see the device file that I will need to use in the configuration kbd file. But I'm using a laptop. Where can  I find the id to use in the configuration?

You could try `xinput list` to get your device name (often it ends with `-kbd`). The FAQ recommends `evtest`.

> And how to use the same setting for the integrated keyboard and the USB keyboard?

There is a new command line parameter for specifying the keyboard you want to use, but unfortunately it's not yet included in any release. For now, you could just copy your kbd file and replace the keyboard device string. I use [script](https://github.com/david-janssen/kmonad/issues/90#issuecomment-700125813) for that.

**Comment by slotThe** [comment]
@andarms Did things work out or do you need further help with this?

**Comment by andarms** [comment]
> @andarms Did things work out or do you need further help with this?

Yes, please. atfer compiling the app using stack I was able to run the kmonad as a command. but, when i try to run with it a configuration file this error is showed
`
Opening KeySink
kmonad: /dev/uinput: openFd: permission denied (Permission denied)
` I already config the uinput user group following the FAQ guide

another question, i'm using evtest   to find the keyboard event of my laptop, the ouput is the following "AT Translated Set 2 keyboard" is that what i need to put on the settings for linux?

**Comment by MaxGyver83** [comment]
> > @andarms Did things work out or do you need further help with this?
> 
> Yes, please. atfer compiling the app using stack I was able to run the kmonad as a command. but, when i try to run with it a configuration file this error is showed
> `Opening KeySink kmonad: /dev/uinput: openFd: permission denied (Permission denied)` I already config the uinput user group following the FAQ guide

I think I have added my user to both groups: `input` and `uinput`. This is my `/etc/udev/rules.d/90-kmonad.rules`:

```c
KERNEL=="uinput", MODE="0660", GROUP="uinput", OPTIONS+="static_node=uinput"
```

> another question, i'm using evtest to find the keyboard event of my laptop, the ouput is the following "AT Translated Set 2 keyboard" is that what i need to put on the settings for linux?

You should put the path to the device file in the KMonad config file, for example `/dev/input/event4`. Usually there is a symbolic link in `/dev/input/by-path/` pointing to this file:

```shell
~ ▶ ll /dev/input/by-path/platform-i8042-serio-0-event-kbd
lrwxrwxrwx 1 root root 9 Jan  5 11:14 /dev/input/by-path/platform-i8042-serio-0-event-kbd -> ../event4
```

I use `/dev/input/by-path/platform-i8042-serio-0-event-kbd` (because I don't know if it's always `event4`).

**Comment by andarms** [comment]
I added my user to input group and still give me the permission denied error but now over the event file.

file permissions:
`
ll /dev/input/by-id/usb-SONiX_Draconic-event-kbd
lrwxrwxrwx 1 root root 9 ene  5 07:56 /dev/input/by-id/usb-SONiX_Draconic-event-kbd -> ../event6
`

kmonad output:
`kmonad: /dev/input/by-id/usb-SONiX_Draconic-event-kbd: openFd: permission denied (Permission denied)
`

**Comment by slotThe** [comment]
Is the module loaded?  What does `modprobe --first-time uinput` return?

**Comment by MuhammedZakir** [comment]
I have encountered a similar issue and had to use `sudo` all this time. Today, by chance, I found this repo [1] which had a solution for this [2]. Remember to reboot your computer.

[1] https://github.com/omakoto/keyremapper-rs#prerequisite

```
# Add self to the input and uinput groups
sudo usermod -aG input $USER 
sudo groupadd uinput
sudo usermod -aG uinput $USER

# See https://github.com/chrippa/ds4drv/issues/93
echo 'KERNEL=="uinput", SUBSYSTEM=="misc", MODE="0660", GROUP="uinput"' | sudo tee /etc/udev/rules.d/90-uinput.rules

# This seems to be needed because uinput isn't compiled as a loadable module these days.
# See https://github.com/chrippa/ds4drv/issues/93#issuecomment-265300511
echo uinput | sudo tee /etc/modules-load.d/uinput.conf
```

Then reboot.

[2] https://github.com/chrippa/ds4drv/issues/93#issuecomment-265300511

**Comment by david-janssen** [comment]
Yeah, I think we need to be a member of `input` to be able to grab the input keyboard, and a member of `uinput` to generate the outputs. Could you verify that you are a member of both groups? When you are and the problem still persist we'll debug further :-).

**Comment by david-janssen** [comment]
Closing was a misclick, soryr.

**Comment by Kabouik** [comment]
This issue is related to mine so I'm posting here in case someone has any information. I'm trying to use `kmonad` in a LXC container running with XWayland. I have added my guest user to the `input` and `uinput` groups (the later didn't exist in the first place) and mounted my `dev/input/` so that they are visible to the guest:

```sh
total 0
drwxr-xr-x 2 root root 100 May 21 02:22 .
drwxr-xr-x 3 root root 220 May 21 02:23 ..
lrwxrwxrwx 1 root root   9 May 21 02:22 platform-1711a000.sound-tavil-event -> ../event4
lrwxrwxrwx 1 root root   9 May 21 02:22 platform-c17a000.i2c-event-kbd -> ../event3
lrwxrwxrwx 1 root root   9 May 21 02:22 platform-soc:gpio_keys-event -> ../event2
```

However, whenever I try to start `kmonad tutorial.kbd` (which I edited to use `/dev/input/by-path/platform-c17a000.i2c-event-kbd` as device file), I get this:

```
kmonad: /dev/uinput: openFd: does not exist (No such file or directory)
```

Whether I run `kmonad` with `sudo` or not. 

Checking `xinput list`, I see that my LXC container is actually not mentioning the device, but instead:

```
user@debian-sid:/$ xinput list
WARNING: running xinput against an Xwayland server. See the xinput man page for details.
⎡ Virtual core pointer                    	id=2	[master pointer  (3)]
⎜   ↳ Virtual core XTEST pointer              	id=4	[slave  pointer  (2)]
⎜   ↳ xwayland-pointer:13                     	id=6	[slave  pointer  (2)]
⎜   ↳ xwayland-relative-pointer:13            	id=7	[slave  pointer  (2)]
⎜   ↳ xwayland-touch:13                       	id=9	[slave  pointer  (2)]
⎣ Virtual core keyboard                   	id=3	[master keyboard (2)]
    ↳ Virtual core XTEST keyboard             	id=5	[slave  keyboard (3)]
    ↳ xwayland-keyboard:13                    	id=8	[slave  keyboard (3)]
```

I am aware that using LXC and Xwayland certainly makes things more complicated and `kmonad` may not work in that corner case, let alone the fact that I do that on a phone and the host is SailfishOS, which is somewhat limited, but maybe someone has encountered and fixed similar issues with `kmonad` in a LXC container?

**Comment by david-janssen** [comment]
I think you have a misunderstanding abount what `/dev/uinput` is pointing at. The `uinput` device actually isn't our input keyboard, it's our output keyboard. `Uinput` is a Linux kernel module that allows for the creation of simulated input devices that can be manipulated programmatically. So, you might either:
- Not be loading the uinput kernel module. (Have you tried: `sudo modprobe uinput`?)
- Have a kernel that doesn't come packaged with this module at all. In which case I'm not sure your problem has a solution.

**Comment by Kabouik** [comment]
Thanks a lot for your answer. Indeed, I think my kernel does not come with this module at all. I'll investigate how feasible it would be to add it, or just drop the idea for that device. In any case, I think I'll be using kmonad on my computers soon; it really looks great.

**Comment by Kabouik** [comment]
Actually it seems `uinput` is there already but not as a module, it's compiled as part of the Android kernel binary (which is the kernel I use for SailfishOS, the host OS for my Debian LXC container). `/dev/uinput` indeed exists and some dev more experienced than me in kernel black magic confirmed that it works (he already used it).

My setup really is uncommon, but he confirmed that theoretically `kmonad` should work on rooted Android devices (same kernel as me, more or less). So I'll continue to investigate. I'm thinking my issue may just be a permission issue with `/dev/input` that may be related to LXC. If anyone reading this has any experience in that stuff, please do ping me!

**Comment by DeepReef11** [comment]
I wanted to say I've found the current doc really hard to follow. This issue helped me but I figured out by deduction. I think the first link in the readme after install doc should be how to make it works (or at beginning of config doc) and should include the snippet above instead of the unclear instruction saying you need to find?? a udev rule and put the path of ?? in input in config (I couldn't find my keyboard properly, using the one that end with kbd in /dev/input/by-path or /dev/input/by-id worked).

**Comment by jokesper** [comment]
I would agree with you, a `--list` option for devices in combination with a bit more than ![image](https://github.com/user-attachments/assets/37d2c00e-4326-4600-aa94-b4ec88650f65) would help a lot.
